decision-records/2024-08-29--dashboard-collision.md

```diff
@@ -0,0 +1,144 @@
+We can update the [best practices doc](https://discourse.charmhub.io/t/grafana-k8s-docs-how-to-create-a-great-charmed-dashboard/14188#use-an-effective-naming-scheme-22),
+and it's up to charm authors to be disciplined about adhering to it. Charm lib will raise if the pattern doesn't match the expected ("test-time linter").
+
+A good option for UIDs could be `f"{charm_name}-{codename}"`, where `codename` is e.g. an abbreviate dashboard title that the charm author needs to make sure is unique within the charm.
+
+Benefits:
+- UID is human readable and sensible.
+
+Disadvantages:
+- Coming up with a codename is toilsome.
+- To adhere to this new standard, would need to modify all existing dashboards.
+
+### (2/P) Automatically overwrite UIDs with a prescribed naming pattern at deploy/relate time, on the provider side
+SHA1 is of length 40 (`len(hashlib.sha1(b"whatever").hexdigest()) == 40`), so calcualte the dashboards UID from hashing (charm_name + rel_path of the dashboard file).
+
+Benefits:
+- The combination of charm_name and rel_path is almost guaranteed to be unique across the ecosystem. Exceptions:
+  - Problem 1a
+  - Local charm's name and dashboard rel_path collide with another charm.
+- Change will be automtically rolled out to everyone by updating the dashboards charm lib.
+
+Disadvantages:
+- Keeping the dashboard itself intact but moving it to another subfolder would break dashboard URL.
+- Would need special attention for aggregation (grafana-agent, cos-proxy) and side-loading (cos-configuration).
+- Overwriting without asking might be surprising for users who want to control UIDs themselves.
+- Dashboards taken from the grafana marketplace would lose their original identity.
+
+### (3/R) Automatically overwrite UIDs with a prescribed naming pattern on relation changed, on the requirer side
+This is possible, thanks to the schema we already have in place (see Appendix 1).
+
+Similar benefits and disadvantages to doing it on the provider side.
+
+### (4/R) On rev-collision, keep only the latest dashboard revision
+
+If we received a "newer dashboard" from a higher revision of a charm, we'd only keep the latest dashboard.
+
+Benefits:
+- A workaround for problem 1a.
+
+Disadvantages:
+- When different charm revisions are related to grafana, will be able to see only the newest dashboard, which may be incompatible with the set of metrics exposed by the older revision.
+
+### (5/R) Place dashboards in folders by charm name
+
+We can create a folder per charm (using the charm name, not the app name) and put the dashboard in there. This is possible because the charm name is already part of the relation data schema (Appendix 1).
+
+### (6/P) Let charm authors or admins specify the folder name under which grafana should place its dashboards
+
+This would solve (2) if charm authors configure their charms correctly. The default behavior would be either the "General" folder (as it is right now) or a folder named after the charm name.
```

Comment:
I think it might be nice from the "not surprising users" perspective to keep the "General" folder as the default. 

I really like the ability to specify a directory though :)